File: 2020/Day03/Haskell/TobogganTrajectory-Part1.hs
import System.IO
import Control.Monad

printArray:: (Show a, Eq a) => [a] -> String
printArray (x:xs)
    | null xs = show x
    | otherwise = show x ++ "\n" ++ printArray xs

findTree:: [[Char]] -> Int -> Int -> Int
findTree world dx dy = findTree_ world dx dy 0 0 0
    where 
        findTree_:: [[Char]] -> Int -> Int -> Int -> Int -> Int -> Int
        findTree_ world dx dy x y total 
            | length world <= y = total
            | (world!!y!!x) == '#' = findTree_ world dx dy (mod (x+dx) (length (head world))) (y+dy) (total+1)
            | otherwise = findTree_ world dx dy (mod (x+dx) (length (head world))) (y+dy) total

main:: IO()
main = do
    handle <- openFile "../input.txt" ReadMode
    contents <- hGetContents handle
    let world = lines contents
    print (findTree world 3 1)

    hClose handle

File: 2020/Day03/Python/TobogganTrajectory-Part1.py
import sys
import collections
import itertools
import numpy as np

world = []
x = y = total = 0

for line in open(f"{sys.path[0]}/../input.txt").read().splitlines():
    world.append(list(line))

while y < len(world):
    x = x % len(world[0])

    if world[y][x] == "#":
        total += 1
    
    x += 3
    y += 1

print(total)